File: sDSA/.Practice Arena/02_BackTracking/nQueen.py
# Solve for 3x3 queen
#todo: Min n is 4;
"""
this is running like 
    ["q1...", 
     "q2...", 
     "q3...",
     "q4..."]
"""
from typing import List

"""
This will give ["....", "....", "....", "...."]
"""
# board Generator
def generateBoard(count: int)-> List[str]:
    board = []
    for _ in range(1,count+1):
        oneRow = ""
        for _ in range(1,count+1):      # or just even do {onerow = '.' * count} !!
            oneRow += "."
        board.append(oneRow)
    return board

def isSafe(board, row, col):
    n = len(board)
    # Check Horizontally
    for j in range(row):
        if board[j][col] == 'Q':
            return False

    # upper Left
    i, j = row - 1, col - 1
    while i >= 0 and j >= 0:
        if board[i][j] == 'Q':
            return False
        i -= 1
        j -= 1

    # Upper Right
    i, j = row - 1, col + 1
    while i >= 0 and j < n:
        if board[i][j] == 'Q':
            return False
        i -= 1
        j += 1

    return True

# backtrack Fuction
def solveQueenBack(board: List[str], solution: List[List[str]], row: int, nQueens: int):
    if row == len(board): # or nQueen
        solution.append(board.copy())
        #? .append(["".join(r) for r in board])
        return
    
    for col in range(nQueens): # NOTE: 0 -> 3
        if isSafe(board, row, col): 
            # this i is traversing throungh the list
            # and col is traversing thrung string[i]
            board[row] = board[row][:col] + "Q" + board[row][col+1:]
            solveQueenBack(board, solution, row+1, nQueens)
            board[row] = board[row][:col] + "." + board[row][col+1:] #todo : Backtrack
    return #todo : Backtrack

"""
solution will change in base case of BackTracking Fuction
"""
#* helperFunction
def solveQueen(nQueens: int)-> List[List[str]]:
    board = generateBoard(nQueens)
    solution = []
    solveQueenBack(board,solution,0, nQueens)
    return solution

#* Main
nQueens = 4
solutions = solveQueen(nQueens)

if 0 == len(solutions):
    print("No sol")
else:
    for s in solutions:
        print(s)

File: sDSA/.Practice Arena/02_BackTracking/nQueen.java
import java.util.List;
import java.util.ArrayList;

class Solution {

    public List<List<String>> solveNQueens(int n) {
        List<List<String>> ans = new ArrayList<>();

        char[][] board = new char[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                board[i][j] = '.';
            }
        }

        boolean[] cols = new boolean[n];
        boolean[] diag1 = new boolean[2 * n]; // row - col + n
        boolean[] diag2 = new boolean[2 * n]; // row + col

        backtrack(0, board, ans, cols, diag1, diag2);
        return ans;
    }

    private void backtrack(int row, char[][] board, List<List<String>> ans,
                           boolean[] cols, boolean[] diag1, boolean[] diag2) {

        if (row == board.length) {
            ans.add(construct(board));
            return;
        }

        for (int col = 0; col < board.length; col++) {
            int d1 = row - col + board.length;
            int d2 = row + col;

            if (cols[col] || diag1[d1] || diag2[d2]) continue;

            board[row][col] = 'Q';
            cols[col] = diag1[d1] = diag2[d2] = true;

            backtrack(row + 1, board, ans, cols, diag1, diag2);

            board[row][col] = '.';
            cols[col] = diag1[d1] = diag2[d2] = false;
        }
    }

    private List<String> construct(char[][] board) {
        List<String> res = new ArrayList<>();
        for (int i = 0; i < board.length; i++) {
            res.add(new String(board[i]));
        }
        return res;
    }
}

public class nQueen {
    public static void main(String[] args) {
        Solution sol = new Solution();

        int n = 4;   // change N here
        List<List<String>> result = sol.solveNQueens(n);

        System.out.println("Total solutions: " + result.size());
        System.out.println();

        int count = 1;
        for (List<String> board : result) {
            System.out.println("Solution " + count++);
            for (String row : board) {
                System.out.println(row);
            }
            System.out.println();
        }
    }
}
